README.md
# Bottom Reveal
An animated Bottom reveal widget, that reveals widgets placed in its back view.

<img src="https://raw.githubusercontent.com/lohanidamodar/flutter_bottom_reveal/master/screenshot/bottomreveal.gif" height="480px"> 

## Usage
To use this plugin, add `bottomreveal` as [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages)

## Example

```dart
import 'package:flutter/material.dart';
import 'package:bottomreveal/bottomreveal.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home:  HomePage(),
    );
  }
}


class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final BottomRevealController _menuController = BottomRevealController();
  @override
  Widget build(BuildContext context){
    return Scaffold(
      appBar: AppBar(
        title: Text('Bottom Reveal Example App'),
      ),
      body: BottomReveal(
        openIcon: Icons.add,
        closeIcon: Icons.close,
        revealWidth: 100,
        revealHeight: 100,
        backColor: Colors.grey.shade900,
        rightContent: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            MaterialButton(
              minWidth: 0,
              child: Icon(Icons.cloud_circle),
              color: Colors.grey.shade200,
              elevation: 0,
              onPressed:() {},
            ),
            MaterialButton(
              minWidth: 0,
              child: Icon(Icons.network_wifi),
              color: Colors.grey.shade200,
              elevation: 0,
              onPressed:() {},
            ),
          ],
        ),
        bottomContent: TextField(
          decoration: InputDecoration(
            filled: true,
            fillColor: Colors.grey,
            contentPadding: const EdgeInsets.all(8.0),
            border: OutlineInputBorder(
              gapPadding: 8.0,
              borderSide: BorderSide(color: Colors.grey),
              borderRadius: BorderRadius.circular(30.0)
            )
          ),
        ),
        controller: _menuController,
        body: ListView.builder(
          padding: const EdgeInsets.all(16.0),
          itemBuilder: (_,index)=>Card(
            child: ListTile(
              title: Text("Item $index"),
              leading: Icon(Icons.cloud_circle),
            ),
          ),
        ),
      ),
    );
  }
}


```


CHANGELOG.md
## 1.0.0

* An animated bottom reveal widget, reveals the background widget on click.


test/bottomreveal_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:bottomreveal/bottomreveal.dart';

void main() {
  const MethodChannel channel = MethodChannel('bottomreveal');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await Bottomreveal.platformVersion, '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:bottomreveal_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/bottomreveal.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;

class BottomRevealController {
  Function close;
  Function open;
  Function toggle;
}

class BottomReveal extends StatefulWidget {
  final BottomRevealController controller;
  final Widget body;
  final Color backColor;
  final Color frontColor;
  final double borderRadius;
  final double revealHeight;
  final double revealWidth;
  final Widget rightContent;
  final Widget bottomContent;
  final IconData openIcon;
  final IconData closeIcon;

  const BottomReveal(
      {Key key,
      this.controller,
      @required this.body,
      this.backColor = Colors.grey,
      this.frontColor = Colors.white,
      this.borderRadius = 50,
      this.revealHeight = 70.0,
      this.revealWidth = 70.0,
      this.rightContent,
      this.bottomContent,
      @required this.openIcon,
      @required this.closeIcon})
      : super(key: key);
  @override
  _BottomRevealState createState() => _BottomRevealState();
}

class _BottomRevealState extends State<BottomReveal>
    with SingleTickerProviderStateMixin {
  AnimationController _controller;
  double bottomMargin;
  double rightMargin;
  bool opened;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 500));
    opened = false;
    bottomMargin = 0;
    rightMargin = 0;
  }

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    widget.controller?.open = _open;
    widget.controller?.close = _close;
    widget.controller?.toggle = _toggle;
    return Scaffold(
      backgroundColor: widget.backColor,
      floatingActionButton: _buildActionButton(),
      body: Stack(
        children: <Widget>[
          if (widget.rightContent != null)
            Align(
              alignment: Alignment.bottomRight,
              child: Container(
                alignment: Alignment.bottomCenter,
                width: widget.revealWidth,
                padding: const EdgeInsets.fromLTRB(8.0, 16.0, 8.0, 90),
                child: widget.rightContent,
              ),
            ),
          if (widget.bottomContent != null)
            Align(
              alignment: Alignment.bottomRight,
              child: Container(
                  padding: EdgeInsets.only(
                      left: 8.0,
                      right: widget.revealWidth,
                      bottom: 8.0,
                      top: 8.0),
                  alignment: Alignment.centerRight,
                  height: widget.revealHeight - 15,
                  child: widget.bottomContent),
            ),
          AnimatedPositioned(
            curve: Curves.easeInOutBack,
            duration: const Duration(milliseconds: 500),
            bottom: bottomMargin,
            right: rightMargin,
            top: -bottomMargin,
            left: -rightMargin,
            child: AnimatedContainer(
              duration: const Duration(milliseconds: 200),
              decoration: BoxDecoration(
                  color: widget.frontColor,
                  borderRadius: opened
                      ? BorderRadius.only(
                          bottomRight: Radius.circular(widget.borderRadius))
                      : null,
                  boxShadow: [
                    BoxShadow(
                      color: widget.backColor.withOpacity(0.5),
                      blurRadius: 5.0,
                    ),
                  ]),
              child: ClipRRect(
                borderRadius: BorderRadius.only(
                    bottomRight: Radius.circular(widget.borderRadius)),
                child: opened && !_controller.isAnimating
                    ? Stack(
                        fit: StackFit.expand,
                        children: <Widget>[
                          widget.body,
                          Container(
                            color: Colors.black12,
                          )
                        ],
                      )
                    : widget.body,
              ),
            ),
          )
        ],
      ),
    );
  }

  FloatingActionButton _buildActionButton() {
    return FloatingActionButton(
      backgroundColor: Colors.pink,
      elevation: 0,
      child: AnimatedBuilder(
        animation: _controller,
        builder: (BuildContext context, Widget child) {
          return Transform(
              alignment: FractionalOffset.center,
              transform: Matrix4.rotationZ(_controller.value * .5 * math.pi),
              child: Icon(_controller.isDismissed
                  ? widget.openIcon
                  : widget.closeIcon));
        },
      ),
      onPressed: () {
        _toggle();
      },
    );
  }

  void _open() {
    setState(() {
      rightMargin = widget.revealWidth;
      bottomMargin = widget.revealHeight;
      _controller.reverse();
      opened = true;
    });
  }

  void _close() {
    setState(() {
      rightMargin = 0;
      bottomMargin = 0;
      _controller.reverse();
      opened = false;
    });
  }

  _toggle() {
    if (opened) {
      _close();
    } else {
      _open();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:bottomreveal/bottomreveal.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final BottomRevealController _menuController = BottomRevealController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Bottom Reveal Example App'),
      ),
      body: BottomReveal(
        openIcon: Icons.add,
        closeIcon: Icons.close,
        revealWidth: 100,
        revealHeight: 100,
        backColor: Colors.grey.shade900,
        rightContent: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            MaterialButton(
              minWidth: 0,
              child: Icon(Icons.cloud_circle),
              color: Colors.grey.shade200,
              elevation: 0,
              onPressed: () {},
            ),
            MaterialButton(
              minWidth: 0,
              child: Icon(Icons.network_wifi),
              color: Colors.grey.shade200,
              elevation: 0,
              onPressed: () {},
            ),
          ],
        ),
        bottomContent: TextField(
          decoration: InputDecoration(
              filled: true,
              fillColor: Colors.grey,
              contentPadding: const EdgeInsets.all(8.0),
              border: OutlineInputBorder(
                  gapPadding: 8.0,
                  borderSide: BorderSide(color: Colors.grey),
                  borderRadius: BorderRadius.circular(30.0))),
        ),
        controller: _menuController,
        body: ListView.builder(
          padding: const EdgeInsets.all(16.0),
          itemBuilder: (_, index) => Card(
            child: ListTile(
              title: Text("Item $index"),
              leading: Icon(Icons.cloud_circle),
            ),
          ),
        ),
      ),
    );
  }
}
